lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+                # Assuming best effort installs so suppress the exception and
+                # mark as a failure UNLESS this is the explicit package.
+                # TODO: Does this interplay properly with Spack's
+                # TODO: 'fail_on_exit' option given the above goal?
+                err = 'Failed to install {0} due to {1}: {2}'
+                tty.error(err.format(pkg.name, exc.__class__.__name__,
+                          str(exc)))
+                self._update_failed(task, True, exc)
+
+                if pkg_id == self.pkg.unique_id:
+                    # TODO: Prefer the following but creates challenges for
+                    # TODO: updating some existing tests.
+                    # reason = 'due to {0}: {1}'.format(exc.__class__.__name__,
+                    #                                   str(exc))
+                    # raise InstallError('Installation of {0} failed {1}'
+                    #                    .format(pkg_id, reason))
+                    raise
+
+            finally:
+                # Remove the install prefix if anything went wrong during
+                # install.
+                if not keep_prefix:
+                    pkg.remove_prefix()
+
+                # The subprocess *may* have removed the build stage. Mark it
+                # not created so that the next time pkg.stage is invoked, we
+                # check the filesystem for it.
+                pkg.stage.created = False
+
+            # Perform basic task cleanup for the installed spec to
+            # include downgrading the write to a read lock
+            self._cleanup_task(pkg, True)
+
+        # Cleanup, which includes releasing all of the read locks
+        self._cleanup_all_tasks()
+
+        # Ensure we properly report if the original/explicit pkg is failed
+        if self.pkg.unique_id in self.failed:
+            msg = ('Installation of {0} failed.  Review log for details'
+                   .format(self.pkg.unique_id))
+            tty.error(msg)
+            raise InstallError(msg)
+
+    install.__doc__ += install_args_docstring
+
+    # Helper method to "smooth" the transition from the
+    # spack.package.PackageBase class
+    @property
+    def spec(self):
+        """The specification associated with the package."""
+        return self.pkg.spec
+
+
+class BuildTask(object):
+    """Class for representing the build task for a package."""
+
+    def __init__(self, pkg, compiler, start, attempts, status, installed):
+        """
+        Instantiate a build task for a package.
+
+        Args:
+            pkg (Package): the package to be installed and built
+            compiler (bool): ``True`` if the task is for a bootstrap compiler,
+                otherwise, ``False``
+            start (int): the initial start time for the package, in seconds
+            attempts (int): the number of attempts to install the package
+            status (int): the installation status
```

Comment:
The doc says `int` but the statuses are strings...

Reply:
Great catch.  (They were in an initial prototype.)  Fixed.